internal/cover/cover.go

```diff
@@ -0,0 +1,254 @@
+		return err
+	}
+
+	if isNewFile {
+		_, err := fmt.Fprintf(f, "mode: %s\n", cover.mode)
+		if err != nil {
+			return err
+		}
+	}
+
+	for i, block := range cover.trackedBlocks {
+		_, err := fmt.Fprintf(f, "%s:%d.%d,%d.%d %d %d\n",
+			toAbsolutePath(block.path),
+			block.start.Line, block.start.Column,
+			block.end.Line, block.end.Column,
+			block.numStmts, dataInts[i+1],
+		)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func dataToInts(data map[string]interface{}) (map[int]int, error) {
+	res := make(map[int]int)
+	for k, v := range data {
+		ki, err := strconv.Atoi(k)
+		if err != nil {
+			return nil, fmt.Errorf("non-int index in data: %s", k)
+		}
+		vf, ok := v.(float64)
+		if !ok {
+			return nil, fmt.Errorf("non-float64 value in data: %v", v)
+		}
+		res[ki] = int(vf)
+	}
+	return res, nil
+}
+
+func (cover *Cover) annotateActions(actions []*ast.Action) []*ast.Action {
+	res := make([]*ast.Action, 0, len(actions))
+	for _, action := range actions {
+		action.Stmts = cover.annotateStmts(action.Stmts)
+		res = append(res, action)
+	}
+	return res
+}
+
+func (cover *Cover) annotateFunctions(functions []*ast.Function) []*ast.Function {
+	res := make([]*ast.Function, 0, len(functions))
+	for _, function := range functions {
+		function.Body = cover.annotateStmts(function.Body)
+		res = append(res, function)
+	}
+	return res
+}
+
+func (cover *Cover) annotateStmtsList(stmtsList []ast.Stmts) []ast.Stmts {
+	res := make([]ast.Stmts, 0, len(stmtsList))
+	for _, stmts := range stmtsList {
+		res = append(res, cover.annotateStmts(stmts))
+	}
+	return res
+}
+
+// annotateStmts takes a list of statements and adds counters to the beginning of
+// each basic block at the top level of that list. For instance, given
+//
+//	S1
+//	if cond {
+//		S2
+//	}
+//	S3
+//
+// counters will be added before S1,S2,S3.
+func (cover *Cover) annotateStmts(stmts ast.Stmts) ast.Stmts {
+	var res ast.Stmts
+	var trackedBlockStmts []ast.Stmt
+	for _, stmt := range stmts {
+		blockEnds := true
+		switch s := stmt.(type) {
+		case *ast.IfStmt:
+			s.Body = cover.annotateStmts(s.Body)
+			s.Else = cover.annotateStmts(s.Else)
+		case *ast.ForStmt:
+			s.Body = cover.annotateStmts(s.Body) // TODO should we do smth with pre & post?
+		case *ast.ForInStmt:
+			s.Body = cover.annotateStmts(s.Body)
+		case *ast.WhileStmt:
+			s.Body = cover.annotateStmts(s.Body)
+		case *ast.DoWhileStmt:
+			s.Body = cover.annotateStmts(s.Body)
+		case *ast.BlockStmt:
+			s.Body = cover.annotateStmts(s.Body)
+		default:
+			blockEnds = false
+		}
+		trackedBlockStmts = append(trackedBlockStmts, stmt)
+		if blockEnds {
+			res = append(res, cover.trackStatement(trackedBlockStmts))
+			res = append(res, trackedBlockStmts...)
+			trackedBlockStmts = []ast.Stmt{}
+		}
+	}
+	if len(trackedBlockStmts) > 0 {
+		res = append(res, cover.trackStatement(trackedBlockStmts))
+		res = append(res, trackedBlockStmts...)
+	}
+	return res
+	// TODO complete handling of if/else/else if
```

Comment:
Let's remove this TODO comment. Seems unlikely we'll handle this case in the near future.